services/user/Supervisor/ui/src/callStack.ts

```diff
@@ -0,0 +1,61 @@
+import { toString } from "@psibase/common-lib";
+import { QualifiedFunctionCallArgs } from "@psibase/common-lib/messaging";
+
+import { OriginationData } from "./utils";
+
+export interface Call {
+    caller: OriginationData;
+    args: QualifiedFunctionCallArgs;
+    startTime?: number;
+}
+
+// Callstack implementation that manages frames for every inter-plugin call
+export class CallStack {
+    private storage: Array<Call> = [];
+
+    push(sender: OriginationData, args: QualifiedFunctionCallArgs): void {
+        console.log(
+            `Callstack: ${" ".repeat(4 * this.storage.length)}+${toString(args)}`,
+        );
+
+        this.storage.push({
+            caller: sender,
+            args: args,
+            startTime: Date.now(),
+        });
+    }
+
+    pop(): Call | undefined {
+        if (this.storage.length === 0) {
+            console.error(`Tried to pop empty callstack`);
+            return undefined;
+        }
```

Comment:
You could use your `isEmpty` method for this.

Reply:
done